src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -185,6 +188,17 @@ def _backchain_one_pass(
                     obj_to_var = dict(
                         zip(ground_op.objects, pnad.op.parameters))
 
+                # For any atom in the necessary_image that wasn't in the
+                # necessary_add_effects, add it as a possible keep effect.
+                for atom in necessary_image - necessary_add_effects:
+                    for obj in atom.objects:
+                        if obj not in obj_to_var:
+                            var, = utils.create_new_variables(
```

Comment:
this is interesting. so with this, would it be possible for keep effects to expand the set of objects in the scope for an NSRT? if so, I'm surprised that you needed to add NotOnTable, because I thought the motivation for that predicate was to deal with scope issues

Reply:
discussed offline -- with NotOnTable, we are assuming keep effects cannot introduce new variables. this simplifies the logic too. cc @NishanthJKumar